### ISSUE TITLE
Should we consolidate py-ncepbufr and bufrlib repos

### ISSUE THREAD

**Issue by srherbener** [JCSDA/py-ncepbufr#23]
The py-ncepbufr repo has recently been updated to use the NCEP bufrlib library version 11.3.0, which happens to match the version that is being used in the bufrlib repo.

It would be nice to track only one copy of the NCEP bufrlib so we have an opportunity to consolidate since py-ncepbufr and bufrlib are on the same version.

Seems like we should discuss this first however.

Leaving the py-ncepbufr and bufrlib repos independent of each other is nice in regard to keeping the two repos independent of each other, however it is not nice in that we would have to track two copies of the NCEP bufrlib code.

The two copies of NCEP bufrlib 11.3.0 are identical between the two repos, except for two new files that were introduced into py-ncepbufr (fortran_open.f and fortran_close.f).

bufrlib repo is a copy (not a fork) of the NCEP bufrlib code with a CMake configuration that Mark O added.

One question is, how open would NCEP be to us contributing our bufrlib repo to them and moving forward with that. I would guess that NCEP already has a repo, but it is not public.

It seems that it would be good to add the two new fortran files to our bufrlib repo, and then attempt to feed that back to NCEP. It would be great if we could clone the NCEP repo when we build, or create a fork now and then to put in specific fixes. This idea of course depends on how NCEP feels about this.

I'm open to other approaches, I just thought the above would be helpful in getting the discussion started.

Thanks!

**Comment by srherbener** [comment]
The system won't let me assign anyone to this issue. I probably don't have the right permissions. Could you assign @srherbener, @markjolah and @jswhit to this for starters. Thanks!

**Comment by markjolah** [comment]
It's important to be able to build NCEP bufrlib with different compilers and options, which was the point of adding the CMake build system.  I am concerned that py-ncepbufr does not provide the same control over the build process, and so cannot be a replacement for JCSDA/bufrlib where Fortan code is involved.   I'm also concerned that py-ncepbufr appears to be Python2 only, which removes motivation to work at unifying the build systems of the two repositories.  I am fine with leaving the versions fixed at 11.3 until there is a good reason to change.  

I would also suggest we look into whether the python 3 package [pybufrkit](https://pybufrkit.readthedocs.io/en/latest/) can replace our need for python/libbufr interfaces.  It is a WMO bufr library, but it is very nice to install and work with.  We would need to understand if the technical differences between bufr formats are actually a factor or not.  Also we would need to develop some benchmarks to test speed.

**Comment by jswhit** [comment]
py-ncepbufr works fine with Python 3

I'm fine with py-ncepbufr using the external bufrlib repo.  I could just keep the extra two fortran files as part of py-ncepbufr.

**Comment by markjolah** [comment]
Yes great that python 3 is supported!  Sorry I was confused between this library and the ecmwf ODB python library which is python2 only.

I suppose the easiest solution would be as suggested for py-ncepbufr to build the two extra files and link against the shared libbufr.so.

I am in the middle of testing some new fortran code with the JCSDA/bufrlib library, which may need some tweeking to the compiler flags.  Once I get that working we can see if we can pass the py-ncepbufr tests when linking against libbufr.so.

**Comment by jswhit** [comment]
See #26.

@srherbener and @markjolah - can you test this?  Note that you need to either build bufrlib with gfortran, or tell numpy.distutils to use ifort (which is what cmake uses by default when building bufrlib).  I did this by setting FC=gfortran before running tools/build.sh.

**Comment by srherbener** [comment]
@jswhit, I've tested the feature/external_bufrlib branch using the library built from the JCSDA/bufrlib repo on my MacBook (GNU Fortran 7.4.0). It built and installed properly and the test/test.py script ran without any messages (meaning the tests passed, correct?). It all looks good. Thanks for making this update!

**Comment by jswhit** [comment]
merged pull request #26.  Still needs to be tested with intel compiler though.

**Comment by srherbener** [comment]
@jswhit, @markjolah, I tried py-ncepbufr using the JCSDA/bufrlib built library using Intel 17.0.1 compilers on Cheyenne, and the test/test.py script runs without any messages and returns a zero. I think that means that the tests pass, and py-ncepbufr works with Intel compilers.

**Comment by jswhit** [comment]
Thanks @srherbener.  Did you have to run `python setup.py build config_fc --fcompiler=intelem`?

**Comment by srherbener** [comment]
No, I have FC=mpiifort in the environment, and that worked with `python setup.py build`.

**Comment by jswhit** [comment]
Hmm  - I bet it still used gfortran for linking then (numpy distutils doesn't use the FC env var). Are you sure it really linked the ifort build bufrlib?
